From 10e4e43bf975ecff4e00fa01a93681ddaf8206ba Mon Sep 17 00:00:00 2001
From: Wang ShaoBo <bobo.shaobowang@huawei.com>
Date: Fri, 26 Feb 2021 20:22:47 +0800
Subject: [PATCH] arm64/mpam: resctrl: Add proper error handling to
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 10e4e43bf975ecff4e00fa01a93681ddaf8206ba

 resctrl_mount()

hulk inclusion
category: bugfix
bugzilla: 48265
CVE: NA

--------------------------------

This function is called only when we mount resctrl sysfs, for error
handling we need to destroy schemata list when next few steps failed
after creation of schemata list.

Fixes: 7e9b5caeefff ("arm64/mpam: resctrl: Add helpers for init and destroy schemata list")
Signed-off-by: Wang ShaoBo <bobo.shaobowang@huawei.com>
Reviewed-by: Cheng Jian <cj.chengjian@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 fs/resctrlfs.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/fs/resctrlfs.c b/fs/resctrlfs.c
index 6fa02741427d..10405e47aafa 100644
--- a/fs/resctrlfs.c
+++ b/fs/resctrlfs.c
@@ -404,11 +404,11 @@ static int resctrl_get_tree(struct fs_context *fc)
 
 	ret = resctrl_id_init();
 	if (ret)
-		goto out;
+		goto out_schema;
 
 	ret = resctrl_group_create_info_dir(resctrl_group_default.kn, &kn_info);
 	if (ret)
-		goto out;
+		goto out_schema;
 
 	if (resctrl_mon_capable) {
 		ret = mongroup_create_dir(resctrl_group_default.kn,
@@ -450,6 +450,8 @@ static int resctrl_get_tree(struct fs_context *fc)
 		kernfs_remove(kn_mongrp);
 out_info:
 	kernfs_remove(kn_info);
+out_schema:
+	schemata_list_destroy();
 out:
 	rdt_last_cmd_clear();
 	mutex_unlock(&resctrl_group_mutex);
-- 
2.26.2

